### ISSUE TITLE
The inferred type exceeds maximum length issue

### ISSUE THREAD

**Issue by hyusetiawan** [sinclairzx81/typebox#273]
I am trying to build a rather complex object and I am not sure what's the best way to compact this, here is the code:
```
export function makeColumnV0_0_1<V extends TSchema, T extends string>(valueSchema: V, versionSchema: T) {
  return Type.Object({
    version: Type.Literal(versionSchema),
    info: Type.String(),
    display: makeExtensibleSchema({
      //   render: OrDisplayObject(makeFunctionWithAPICell(valueSchema, DisplaySchema, Type.Never())),
      render: objectOrCellAPI(valueSchema, DisplaySchema),
    }),
    parse: makeExtensibleSchema({
      logic: makeFunctionWithAPICell(valueSchema, valueSchema, Type.Any()),
    }),
    filters: RecordString(makeFilterSchema(valueSchema)),
    config: RecordString(makeConfigSchema(valueSchema)),
    events: Maybe(makeEventsSchema(valueSchema)),
    menu: Maybe(
      Type.Object({
        column: Maybe(makeColumnMenuSchema(valueSchema)),
        cell: Maybe(makeCellMenuSchema(valueSchema)),
      }),
    ),
    value: Type.Union([
      Type.Object({
        type: Type.Literal('cell'),
        form: objectOrCellAPI(valueSchema, DisplayInputSchema),
      }),
      Type.Object({
        type: Type.Literal('url'),
        method: Type.Union([Type.Literal('post'), Type.Literal('get')]),
        params: Maybe(RecordString(Type.String())), // can refer to other columns, for use case of SKU in one column -> price column by SKU
        headers: Maybe(RecordString(Type.String())), // can refer to other columns, for use case of SKU in one column -> price column by SKU
      }),
    ]),
  })
}

const ColumnSchemaNumber0_0_1 = makeColumnV0_0_1(Type.Number(), 'number.0.0.1')
const ColumnSchemaString0_0_1 = makeColumnV0_0_1(Type.String(), 'string.0.0.1')
const ColumnSchemaBoolean0_0_1 = makeColumnV0_0_1(Type.Boolean(), 'boolean.0.0.1')
const ColumnSchemaDate0_0_1 = makeColumnV0_0_1(Type.Number(), 'date.0.0.1')

export const ColumnSchema = Type.Union([
  makeColumnV0_0_1(Type.Number(), 'number.0.0.1'),
  makeColumnV0_0_1(Type.String(), 'string.0.0.1'),
  makeColumnV0_0_1(Type.Boolean(), 'boolean.0.0.1'),
  makeColumnV0_0_1(Type.Number(), 'date.0.0.1'),
  makeColumnV0_0_1(Type.Array(Type.Number()), 'number[].0.0.1'),
])
```
The problem starts when I add `  makeColumnV0_0_1(Type.Array(Type.Number()), 'number[].0.0.1'),` but i think i am just really hitting the threshold because as I remove the properties of the Column (ex: remove `display` or `value`) it would hit a point where it would not error out. Given the structure, is there a way to optimize these definitions or structure in a more scalable way? 
It boils down to the `Column` object having the same structure, just a different "version" and with that different values, it can contain eg: Type.Number/Type.String/etc

**Comment by sinclairzx81** [comment]
@hyusetiawan Hi. It's difficult for me to see what's going on without having the type factories (`objectOrCellAPI`, `makeExtensibleSchema`, `Maybe`, `makeColumnMenuSchema`, etc) as I can't repro the code and see the actual error. It's usually helpful to include a https://www.typescriptlang.org/play link in these cases.

Overall, the approach you're using looks ok, and using generic functions to compute common types with dependent inner types is supported, so would need to reproduce the error first before advising next steps. My initial thoughts are that I don't think there would be anything TypeBox could reasonably do to solve hard limits in TypeScript's inference capabilities internal to the library, but there may alternate ways to represent the type using `Type.Unsafe<T>` to by-pass hard limits (assuming there's not a better way to express the current type). Note this is the first time I've seen TS complain about inferred type length since TB optimizations were implemented a few versions ago, so again, very interested to see the repro.

If you can provide a repro via TS playground link, I'm happy to have a deeper look.
Cheers
S

**Comment by sinclairzx81** [comment]
See the link below for a fairly complex type that attempts to replicate some of the type composition in your example. I can't seem to produce a type that would push inferred length limits, but just to confirm you're on the latest version of TypeScript?

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIY64ykzD5wC+cAZlBCHAOQACAzsAHb4ANqWBQA9DHQYARhAAeHALAAoVePFwAahnwxoAZgC0Q4AGsMcAO7AYhOABMMmfk-7xgDuDxhQBAc1V8CH4fOABBOABeOAAeXDgMeRgMVx5vXwCAPgAKADdSIQBXDAAuOFwASmislGkAOgB5GQArXRgchFU4OE9ytEx6gBlbDChC-MKSyqxuuHl+hoA5IpAZMZyZudRFwZW1ja2VHoAvXYx6-fWoTdV6StV1TSHSKH9LMFZMWFRvDHgIExtO1oCZzBgeEEQmEAELROqDZptPSdOakACM5XCOQ44Q4MzgMkxERxMPxOHwxOxHAAwuS0QAmLE4gAi5MJTJJHAAouz8JzqQAxenHOCkAzMjgAcXZMglXIAEnz5dSAJIinqkAAskoAUrKdVyANJ8w3UoYasUAVklAFlZTauUs+Y7qY1LaQAGySgAKsu9XIAinyA9SAEoegDsktwsujXOQfPj1IAqh6ABySrSyzNcgDqfNz1IAGh6AJySgCasorXIAWnza9S8Ud7o8VBo4ILSHpoL8mEVBDQQnAcqR+IUIP4IEV0vo4PgoBhqBgaRBiiB+A8VAOh8ARzSlyvYshEslUg50sgCMQyBQzyk0hk-Px-LkeEQSKR+jgCsUyhU1RdKKS4wEUUD8AiFxIu0qKij0UiYOcwyjOMQiTP+Rw9D0f4lOUH63qQcxtio9Dtp2Kb8PukGIZYwShL4RR6NRo42HYcBLvg4F8CE270WEp4xAMFxhro3HAHkGA5IJtTCU0rSwcBPQ8BgQhMMh4RQOMqDSVhcCDtRyGUdROQANpzD0h7LikOQwjguKWpZR42XZnBknpTnWVJrm0o5cBWSutn2WyHn+c53n2byoUBS59nCtF4VBZwMoJV5SUcEqqWBT56pZbFnD6nlEWcCaekALrVHclTbk8cC0b06QQGYqj1YJFRUDQ+CxLRgIoFk7ZAA)

```typescript
// computed output type should omit excessive property expansion
type T = {
    self: ...[]; // no expansion on recursive types.
    union: {
        type: "A";
        value: {
            a1: {
                id: "A";
                x: number;
                y: number;
                z: number;
            };
            b1: {
                id: "B";
                x: number;
                y: number;
                z: number;
            };
            c1: {
                id: "C";
                x: number;
                y: number;
                z: number;
            };
            a2: {
                id: "D";
                x: number;
                y: number;
                z: number;
            };
            ... 22 more ...;  // excessive properties should omit
            c9: {
                id: "A";
                x: number;
                y: number;
                z: number;
            };
        };
    } | ... 9 more ... | { // excessive unions should omit
        ...;
    };
}
```

**Comment by hyusetiawan** [comment]
Yeah apologies i didn't share all of the code, I'm afraid it'd be too much
so i just shared the structure, but you're able to masterfully capture the
intent though!

The Recursive solution works, i ended up doing:

Type.Recursive(T => Type.Union([makeColumn...

I wonder what's the logic behind it? It seems by simply wrapping the
original Type.Union that i have above in a Recursive call made it work.

**Comment by sinclairzx81** [comment]
@hyusetiawan All good :) Sounds like the that solved things so will close off this issue.

Just on recursive types, TypeBox currently implements these via an HKT (higher-kinded-type) that allows TS to evaluate types lazily on property navigation (or when checking a finite sized data structure). This helps prevent TS from infinitely expanding on the recursive type, where expansion only happens on property navigation. The inference logic for recursive types is quite complicated (and a work in progress), future implementations may be able to implement this via TS instantiation expressions that were introduced in 4.7 (iirc). But have a read up on "TS higher kinded types" and "Instantiation Expressions" if you're interested in a bit more information.

As for the schema, the recursive parameter `TSelf` is just `{ $ref: '...' }` which is auto referenced to the outer schemas generated `$id`.

Hope this helps!
Cheers!
S

**Comment by hyusetiawan** [comment]
I found another issue, the `Type.Recursive` works if the schema instance is not exported, upon `export` it will trigger the same exceeds maximum length issue, unfortunately.

**Comment by ThyDevourer** [comment]
Sorry for reviving this closed issue, I'm trying to create a generic function that takes a Type.Object schema and returns another, bigger schema with the passed argument as one of its keys, and another generic function to extend that same initial schema (with Type.Intersect). If i nest the function calls like this:
```
makeResponseSchema(addIdToSchema(schema))
```
i get the following warning in my IDE:
`Type instantiation is excessively deep and possibly infinite.`

Heres a minimal reproduction of the issue: https://replit.com/@ThyDevourer/ComplexDarkredPup#index.ts

Strangely, if I first save the result of the inner function call in a variable and then pass that variable to the outer function (as shown in the replit's index2.ts), i don't get any errors or warnings

**Comment by sinclairzx81** [comment]
@ThyDevourer Hi!

Hey, this is a excellent find! I've had a look into this and it appears to be a tightening of TS inference depth from TS v4.7.x onwards. 

The following works ok on TS 4.6.4.

[TypeScript Link Here](https://www.typescriptlang.org/play?ts=4.6.4#code/JYWwDg9gTgLgBAbzgFQPICMBWBTAxjAGhQE8xsiBlGAQxmFzgF84AzKCEOAcgAEBnYADtcAG2rAoAehils6CAA8uAWABQa3BEF94EYABMKuABbYQ1OAF4SZAHQYc+ABQI1cOAH0DALhvZbVFBCAOZOAJRqjERI1Pr6wHRa1CIACuxksMDYfL4syXzYTBHqqpra8LH6qAbIEEam5lZwADzI9WYW2Aow2IL6fCgOeDAAfE58Jh2+bZPmYVYjfrYAkoI9UAXOANpu7nATDdREu+56hrPUagC6xRpaOnAArgVQ7Y3WyLL2WMNOu66qPZwQT0ADW0y+gRC4QIJzgHWAIghdihglCYVhgKYmPcMTiCWASVS6WwmWyuXyhUYalupXu8HMoOwACVspBtNg3hZrK0ufDur1+oMfvgxrt9LRqFzplyaQtdp87ENnEgJTRfGqpRdsXC8fFEoJkmkIBk6OTWJSmLtaXdynAoGz7pztdZGSzHRyuU5KtV9LUvc9SVywrSZGQUE0qLR6M0w9gICx7R6ClzFpJJCgKHAACy2ABstmzcAAtIsIKCAIRAA)

```typescript
import { TObject, Type, Static } from '@sinclair/typebox'

const oidSchema = Type.Object({
  _id: Type.String()
}, { additionalProperties: false })

const addOidToSchema = <TSchema extends TObject>(schema: TSchema) => Type.Intersect([
    schema, 
    oidSchema
])

const userSchema = Type.Object(
  {
    nick: Type.String(),
    email: Type.String(),
  },
  { additionalProperties: false }
)

const makeResponseSchema = <TSchema extends TObject>(
  dataSchema: TSchema
) =>
  Type.Object({ data: dataSchema },
    { additionalProperties: false }
  )


const responseSchema = makeResponseSchema(addOidToSchema(userSchema))

type T = Static<typeof responseSchema> // TS 4.6.4 -> ok!
```
I'll take a look at this over the next few days and report back if I can find a way to have TS evaluate for this case, however I don't anticipate a easy resolution for this while supporting older versions of the compiler (I suspect these constraints are largely a result of the new TS [instantiation expressions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions) that were added in 4.7).

As you've noticed, pulling `addOidToSchema(userSchema)` into it's own variable is a way to have TS evaluate the type early (vs TS having to evaluate everything on `makeResponseSchema(addOidToSchema(userSchema))`. Although not ideal, it is a "trick" you can use to keep inference fairly flat, I would recommend this as a workaround for the time being.

```typescript
const userSchemaWithOid = addOidToSchema(userSchema)

const responseSchema = makeResponseSchema(userSchemaWithOid)
```
Thanks for the heads up on this, I hope the workaround helps in the interim.
Cheers
S

**Comment by sinclairzx81** [comment]
@ThyDevourer Hey, just as a quick addendum, it is also possible to have TS evaluate early through `as` assertion


### Instantiation Expression
```typescript
const responseSchema = makeResponseSchema(
   addOidToSchema(userSchema) as ReturnType<typeof addOidToSchema<typeof userSchema>>
)
```
### Explicit

```typescript
type Evaluated<TSchema extends TObject> = TIntersect<[TSchema, typeof oidSchema]>

const responseSchema = makeResponseSchema(
  addOidToSchema(userSchema) as Evaluated<typeof userSchema>
)
```
Again, ill do a bit of digging, but don't anticipate a trivial solution under the current TypeBox version constraints. Would be open to some experimentation in eager type evaluation in TypeScript. I think it would be good for TB to have some feature in place to allow for the composition of complex generic types where inference limits may pop out. Initial thoughts would be some form of utility type (some generic form of `Evaluated` above), but certainly open to suggestions if you've got some thoughts around this.

Cheers!

**Comment by sinclairzx81** [comment]
@ThyDevourer Here's another thing you can do to avoid prespecifying the generic type.

### Runtime Evaluate with Array
```typescript
const responseSchema = makeResponseSchema([addOidToSchema(userSchema)][0])
```

**Comment by sinclairzx81** [comment]
@ThyDevourer Hi, just as a quick follow up. This issue has been properly resolved on `0.26.0` with the generalization of `TIntersect`. The following code has been tested from TS versions `4.1.5` > `5.0.2`. The only change is replacing the `TObject` generic constraint with a general constraint of `TSchema`.

[TypeScript Link Here](https://www.typescriptlang.org/play?ts=5.0.2#code/JYWwDg9gTgLgBAbzgFQMoGMAWBTEBDAGhQE8xsjUY8Zh04BfOAMyghDgHIABAZ2ADt0AGzzAoAehilsAIwgAPDgFgAUKvQR+PeBGAATDDnxwAvCTIA6APIyAVtnQwAFAlVw4AfX0Auc9guUUAIA5k4AlKr0REh4enrANJp4QgAKrGSwwNg8vkzJPNgMEWoqGlrwsXpW+sgQhrh4pnAAPMhw2PIw2Px6PCj1+AB8TjxYDb7IYaaDfhYAkvxdUAWOTgDabu5wo0aEcJvuugZj+KoAusXqmtpwAK4FUAONZsjS1nYOzpuuKltw-LQANYTN6BELhAgHdr4YBCEGWMH8UJhSG-Bio9wxOIJYBJVLpbCZbK5fKFeiqS6la7wfCA7AAJWykC02CeTVa7U63V6-ROeGGmz01DwTwmFOmm1elhs9lWSCFVF8CpFfPRUKx8US-GSaQgGRoxOYpIYm0pV3KcCgTOurNVZlpDOtLKeTkq1T0tRd90JTzClKkZBQTUo1FozQD2AgTEtToKTxm4nEcAjQbMPz+GczWezOcTW2VvnTOeLJezeb+APQwO2MCCSIA3FDS83i+Wtg1Yb5tHXgo20S2B2Wkww4AAyRBNwdTuBtzw+Gs9vvT5fucv0Jcr6d5+hAA)

```typescript
import { TSchema, Type, Static } from '@sinclair/typebox'

const oidSchema = Type.Object({
  _id: Type.String()
}, { additionalProperties: false })

const addOidToSchema = <T extends TSchema>(schema: T) => Type.Intersect([
    schema, 
    oidSchema
])

const userSchema = Type.Object(
  {
    nick: Type.String(),
    email: Type.String(),
  },
  { additionalProperties: false }
)

const makeResponseSchema = <T extends TSchema>(
  dataSchema: T
) =>
  Type.Object({ data: dataSchema },
    { additionalProperties: false }
  )


const responseSchema = makeResponseSchema(addOidToSchema(userSchema))

type T = Static<typeof responseSchema> // type T = {
                                       //   data: {
                                       //     nick: string;
                                       //     email: string;
                                       //  } & {
                                       //    _id: string;
                                       //  };
                                       // }
```
